waves-ext/src/main/scala/com/wavesplatform/dex/grpc/integration/services/WavesBlockchainApiGrpcService.scala

```diff
@@ -2,79 +2,135 @@ package com.wavesplatform.dex.grpc.integration.services
+
     // https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
-    val metadata = new Metadata()
-    metadata.put(descKey, "Shutting down")
+    val metadata      = new Metadata(); metadata.put(descKey, "Shutting down")
     val shutdownError = new StatusRuntimeException(Status.UNAVAILABLE, metadata) // Because it should try to connect to other DEX Extension
 
     realTimeBalanceChangesSubscribers.forEach { _.onError(shutdownError) }
     realTimeBalanceChangesSubscribers.clear()
   }
 
+  private def getAddressesChangedAssets(transactionStateUpdates: Seq[StateUpdate]): Map[Address, Set[Asset]] = {
+    transactionStateUpdates.foldLeft(Map.empty[Address, Set[Asset]]) {
+      case (result, stateUpdate) => result |+| stateUpdate.balances.groupBy(_._1).view.mapValues(_.map(_._2).toSet).toMap
+    }
+  }
+
+  private def getAddressesChangedAssets(diff: Diff): Map[Address, Set[Asset]] = diff.portfolios.view.mapValues(_.assetIds).toMap
+
+  private val txDiffs                                                                 = new ConcurrentHashMap[Transaction, Diff]()
+  private val currentMaxHeight: AtomicInteger                                         = new AtomicInteger(context.blockchain.height)
+  private val isRollback: AtomicBoolean                                               = new AtomicBoolean(false)
+  private val storingChangesDuringRollback: AtomicReference[Map[Address, Set[Asset]]] = new AtomicReference(Map.empty)
+
+  private val blockchainBalanceUpdates: Observable[Map[Address, Set[Asset]]] = context.blockchainUpdated.map {
+    case BlockAppended(_, newHeight, _, _, _, transactionStateUpdates) =>
+      val changes = getAddressesChangedAssets(transactionStateUpdates)
+      if (isRollback.get()) {
+        storingChangesDuringRollback.updateAndGet(_ |+| changes)
+        if (newHeight != currentMaxHeight.get) Map.empty[Address, Set[Asset]]
+        else {
+          isRollback.set(false)
+          storingChangesDuringRollback.getAndSet(Map.empty)
+        }
+      } else {
+        currentMaxHeight.set(newHeight)
+        changes
+      }
+
+    case MicroBlockAppended(_, _, _, _, transactionStateUpdates) => getAddressesChangedAssets(transactionStateUpdates)
+    case MicroBlockRollbackCompleted(_, _)                       => Map.empty[Address, Set[Asset]]
+    case RollbackCompleted(_, _)                                 => isRollback.set(true); Map.empty[Address, Set[Asset]]
+  }
+
+  private val utxBalanceUpdates: Observable[Map[Address, Set[Asset]]] = context.utxEvents.map {
+    case TxAdded(tx, diff) =>
+      tx match {
+        case et: ExchangeTransaction if ignoredExchangeTxSenderPublicKey.contains(et.sender.toString) => Map.empty[Address, Set[Asset]]
+        case otherTx =>
+          txDiffs.putIfAbsent(otherTx, diff)
+          getAddressesChangedAssets(diff)
+      }
+    case TxRemoved(tx, _) => Option(txDiffs remove tx).map(getAddressesChangedAssets).getOrElse(Map.empty)
```

Comment:
We can store `getAddressesChangedAssets(diff)` in `txDiffs` instead of `diff`. Then we won't need to recalculate it in `TxRemoved`

Reply:
Done